File: python/leetcode/131.py
class Solution(object):
    def partition(self, s):
        if not s: return []
        res = [] 
        self.help(s, [], res)
        return res

    def help(self, s, ans, res):
        if not s:
            res.append(ans)
            return 
        for i in range(1, len(s) + 1):
            if s[:i] == s[:i][::-1]:
                self.help(s[i:], ans + [s[:i]], res)
        

File: cpp/leetcode/131.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> ans;
        dfs(0, s, ans, res);
        return res;
    }
    void dfs(int start, string& s, vector<string>& ans, vector<vector<string>>& res){
        if(start == s.size()){
            res.push_back(ans);
            return;
        }
        for(int i = start; i < s.size(); ++i){
            int l = start, r = i;
            while(l < r && s[l] == s[r]) {l++; r--;}
            if(l >= r){
                ans.pop_back(s.substr(start, i - start + 1)) // i included 
                dfs(i+1, s, ans, res); // start from i + 1
                ans.pop_back();
            }
        }
    }
};